File: backend/src/main/java/com/leavemarker/service/HolidayService.java
package com.leavemarker.service;

import com.leavemarker.dto.holiday.HolidayRequest;
import com.leavemarker.dto.holiday.HolidayResponse;
import com.leavemarker.entity.Company;
import com.leavemarker.entity.Holiday;
import com.leavemarker.exception.BadRequestException;
import com.leavemarker.exception.ResourceNotFoundException;
import com.leavemarker.repository.CompanyRepository;
import com.leavemarker.repository.HolidayRepository;
import com.leavemarker.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class HolidayService {

    private final HolidayRepository holidayRepository;
    private final CompanyRepository companyRepository;

    @Transactional
    public HolidayResponse createHoliday(HolidayRequest request, UserPrincipal currentUser) {
        Company company = companyRepository.findById(currentUser.getCompanyId())
                .orElseThrow(() -> new ResourceNotFoundException("Company not found"));

        if (request.getDate().isBefore(LocalDate.now())) {
            throw new BadRequestException("Cannot create holiday for past dates");
        }

        Holiday holiday = Holiday.builder()
                .company(company)
                .name(request.getName())
                .date(request.getDate())
                .type(request.getType())
                .state(request.getState())
                .active(request.getActive())
                .build();

        holiday = holidayRepository.save(holiday);
        return mapToResponse(holiday);
    }

    public HolidayResponse getHoliday(Long id, UserPrincipal currentUser) {
        Holiday holiday = holidayRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Holiday not found"));

        if (!holiday.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        return mapToResponse(holiday);
    }

    public List<HolidayResponse> getAllHolidays(UserPrincipal currentUser) {
        List<Holiday> holidays = holidayRepository.findByCompanyIdAndDeletedFalse(
                currentUser.getCompanyId());
        return holidays.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public List<HolidayResponse> getActiveHolidays(UserPrincipal currentUser) {
        List<Holiday> holidays = holidayRepository.findByCompanyIdAndActiveAndDeletedFalse(
                currentUser.getCompanyId(), true);
        return holidays.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public List<HolidayResponse> getHolidaysByDateRange(
            LocalDate startDate, LocalDate endDate, UserPrincipal currentUser) {
        List<Holiday> holidays = holidayRepository.findByCompanyIdAndDateRange(
                currentUser.getCompanyId(), startDate, endDate);
        return holidays.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public HolidayResponse updateHoliday(Long id, HolidayRequest request, UserPrincipal currentUser) {
        Holiday holiday = holidayRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Holiday not found"));

        if (!holiday.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        holiday.setName(request.getName());
        holiday.setDate(request.getDate());
        holiday.setType(request.getType());
        holiday.setState(request.getState());
        holiday.setActive(request.getActive());

        holiday = holidayRepository.save(holiday);
        return mapToResponse(holiday);
    }

    @Transactional
    public void deleteHoliday(Long id, UserPrincipal currentUser) {
        Holiday holiday = holidayRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Holiday not found"));

        if (!holiday.getCompany().getId().equals(currentUser.getCompanyId())) {
            throw new BadRequestException("Access denied");
        }

        holiday.setDeleted(true);
        holidayRepository.save(holiday);
    }

    private HolidayResponse mapToResponse(Holiday holiday) {
        return HolidayResponse.builder()
                .id(holiday.getId())
                .name(holiday.getName())
                .date(holiday.getDate())
                .type(holiday.getType())
                .state(holiday.getState())
                .active(holiday.getActive())
                .companyId(holiday.getCompany().getId())
                .companyName(holiday.getCompany().getName())
                .build();
    }
}

File: backend-go/internal/service/holiday_service.go
package service

import (
	"time"

	"github.com/leavemarker/backend-go/internal/dto"
	"github.com/leavemarker/backend-go/internal/models"
	"github.com/leavemarker/backend-go/internal/repository"
	"github.com/leavemarker/backend-go/internal/utils"
)

// HolidayService handles holiday business logic
type HolidayService struct {
	holidayRepo           *repository.HolidayRepository
	planValidationService *PlanValidationService
}

// NewHolidayService creates a new HolidayService
func NewHolidayService(
	holidayRepo *repository.HolidayRepository,
	planValidationService *PlanValidationService,
) *HolidayService {
	return &HolidayService{
		holidayRepo:           holidayRepo,
		planValidationService: planValidationService,
	}
}

// CreateHoliday creates a new holiday
func (s *HolidayService) CreateHoliday(companyID uint, req dto.HolidayRequest) (*dto.HolidayResponse, error) {
	// Validate plan limit
	if err := s.planValidationService.ValidateHolidayLimit(companyID); err != nil {
		return nil, err
	}

	holiday := &models.Holiday{
		CompanyID: companyID,
		Name:      req.Name,
		Date:      req.Date,
		Type:      req.Type,
		State:     req.State,
		Active:    req.Active,
	}

	if err := s.holidayRepo.Create(holiday); err != nil {
		return nil, utils.NewInternalServerError("Failed to create holiday")
	}

	return s.toHolidayResponse(holiday), nil
}

// GetHoliday gets a holiday by ID
func (s *HolidayService) GetHoliday(id, companyID uint) (*dto.HolidayResponse, error) {
	holiday, err := s.holidayRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Holiday not found")
	}

	if holiday.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	return s.toHolidayResponse(holiday), nil
}

// GetAllHolidays gets all holidays for a company
func (s *HolidayService) GetAllHolidays(companyID uint) ([]dto.HolidayResponse, error) {
	holidays, err := s.holidayRepo.FindByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get holidays")
	}

	responses := make([]dto.HolidayResponse, len(holidays))
	for i, holiday := range holidays {
		responses[i] = *s.toHolidayResponse(&holiday)
	}

	return responses, nil
}

// GetActiveHolidays gets all active holidays for a company
func (s *HolidayService) GetActiveHolidays(companyID uint) ([]dto.HolidayResponse, error) {
	holidays, err := s.holidayRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get holidays")
	}

	responses := make([]dto.HolidayResponse, len(holidays))
	for i, holiday := range holidays {
		responses[i] = *s.toHolidayResponse(&holiday)
	}

	return responses, nil
}

// GetHolidaysByDateRange gets holidays by date range for a company
func (s *HolidayService) GetHolidaysByDateRange(companyID uint, startDate, endDate time.Time) ([]dto.HolidayResponse, error) {
	holidays, err := s.holidayRepo.FindByCompanyIDAndDateRange(companyID, startDate, endDate)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get holidays")
	}

	responses := make([]dto.HolidayResponse, len(holidays))
	for i, holiday := range holidays {
		responses[i] = *s.toHolidayResponse(&holiday)
	}

	return responses, nil
}

// UpdateHoliday updates a holiday
func (s *HolidayService) UpdateHoliday(id, companyID uint, req dto.HolidayRequest) (*dto.HolidayResponse, error) {
	holiday, err := s.holidayRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Holiday not found")
	}

	if holiday.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	holiday.Name = req.Name
	holiday.Date = req.Date
	holiday.Type = req.Type
	holiday.State = req.State
	holiday.Active = req.Active

	if err := s.holidayRepo.Update(holiday); err != nil {
		return nil, utils.NewInternalServerError("Failed to update holiday")
	}

	return s.toHolidayResponse(holiday), nil
}

// DeleteHoliday soft deletes a holiday
func (s *HolidayService) DeleteHoliday(id, companyID uint) error {
	holiday, err := s.holidayRepo.FindByID(id)
	if err != nil {
		return utils.NewNotFoundError("Holiday not found")
	}

	if holiday.CompanyID != companyID {
		return utils.NewForbiddenError("Access denied")
	}

	return s.holidayRepo.Delete(id)
}

func (s *HolidayService) toHolidayResponse(holiday *models.Holiday) *dto.HolidayResponse {
	return &dto.HolidayResponse{
		ID:        holiday.ID,
		Name:      holiday.Name,
		Date:      holiday.Date,
		Type:      holiday.Type,
		State:     holiday.State,
		Active:    holiday.Active,
		CreatedAt: holiday.CreatedAt,
		UpdatedAt: holiday.UpdatedAt,
	}
}
